### PR TITLE
COR-1990 Refactoring customers sync flows

### PR DESCRIPTION
**Pull request by DanielxC**
**Moved to #93** 

## PR Description ##
This PR is for refactoring customers sync flows (cron + customerSaveAfter + customerAddressUpdate), for shared logic, readability and performance, this PR contain mostly renaming, extracting logic to shared private methods, reducing uneeded code, and adding logs entries and logs messages explanations.

* extracted creating customers collection with Yotpo data query, to its on private method `createCustomerCollectionWithYotpoSyncDataQuery`.

## Developer Notes ##
This PR doesn't contains any bug fixes, as the logic remained the same.

### CODE REVIEW

File: Model/Sync/Customers/Main.php [hunk 1]
```diff
@@ -97,14 +97,6 @@ public function createCustomerSyncData($response, $magentoCustomerId)
...
-        $finalData = [];
-        $finalData[] = [
-            'customer_id'        =>  $customerSyncData['customer_id'],
-            'synced_to_yotpo'    =>  $customerSyncData['synced_to_yotpo'],
-            'response_code'      =>  $customerSyncData['response_code'],
-            'store_id'           =>  $customerSyncData['store_id'],
-            'sync_status'        =>  $customerSyncData['sync_status']
-        ];
-        $this->insertOnDuplicate('yotpo_customers_sync', $finalData);
+        $this->insertOnDuplicate('yotpo_customers_sync', $customerSyncData);
```
Comment (LiorGingi) [inline Model/Sync/Customers/Main.php:100]: Are you sure that the data that is being saved in this section is exactly the same as in the original implementation of the method?

Comment (LiorGingi) [reply Model/Sync/Customers/Main.php:100]:
```suggestion
        $this->insertOnDuplicate('yotpo_customers_sync', [$customerSyncData]);
```

The callers of this method shouldn't be aware of the fact that this should be passed as array.

File: Model/Sync/Customers/Processor.php [hunk 2]
```diff
@@ -440,4 +428,20 @@ public function retryCustomersSync()
             echo 'No customer data to process.' . PHP_EOL;
         }
     }
+
+    /**
+     * @param array $customerSyncData
+     * @param string $currentTime
+     * @return void
+     */
+    private function updateCustomerSyncData(array $customerSyncData, $currentTime)
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php:458]: I think you should separate into 2 methods:
* One that prepares the data object.
* The second that upserts the data to the DB table (which in this case, is `insertOrUpdateCustomerSyncData`).

File: Model/Sync/Customers/Processor.php [hunk 3]
```diff
@@ -229,35 +229,14 @@ public function processEntities($retryCustomerIds = [])
...
-            $customerCollection->getSelect()
-                ->where('e.website_id = ? ', $websiteId);
-        }
-        $customerCollection->getSelect()->limit($batchSize);
+        $customerCollectionWithYotpoDataQuery =
+            $this->createCustomerCollectionWithYotpoSyncDataQuery($storeId, $retryCustomerIds, $customerAccountShared, $websiteId, $batchSize);
         $magentoCustomers = [];
-        foreach ($customerCollection->getItems() as $customer) {
-            $id = $customer->getId();
+        foreach ($customerCollectionWithYotpoDataQuery->getItems() as $customerWithYotpoData) {
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php]: customerWithYotpo**Sync**Data

File: Model/Sync/Customers/Processor.php [hunk 4]
```diff
@@ -232,13 +232,24 @@ public function processEntities($retryCustomerIds = [])
...
             $this->createCustomerCollectionWithYotpoSyncDataQuery($storeId, $retryCustomerIds, $customerAccountShared, $websiteId, $batchSize);
         $magentoCustomers = [];
-        foreach ($customerCollectionWithYotpoDataQuery->getItems() as $customerWithYotpoData) {
-            $id = $customerWithYotpoData->getId();
-            $id = explode('-', $id);
-            $customerWithYotpoData->setId($id[0]);
-            $magentoCustomers[$customerWithYotpoData->getId()] = $customerWithYotpoData;
-        }
-        if ($magentoCustomers) {
+        $customerCollectionWithYotpoData = $customerCollectionWithYotpoDataQuery->getItems();
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php:248]: customerCollectionWithYotpo**Sync**Data

File: Model/Sync/Customers/Processor.php [hunk 5]
```diff
@@ -232,13 +232,24 @@ public function processEntities($retryCustomerIds = [])
...
+            $this->yotpoSmsBumpLogger->info(
+                __(
+                    'No customers that should be synced found - Magento Store ID: %1, Name: %2',
+                    $storeId,
+                    $this->config->getStoreName($storeId)
+                ),
+                []
+            );
+        } else {
+            foreach ($customerCollectionWithYotpoData as $customerWithYotpoData) {
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php]: customerWithYotpo**Sync**Data

File: Model/Sync/Customers/Processor.php [hunk 6]
```diff
@@ -372,7 +372,7 @@ public function resetCustomerSyncStatus($customerId, $customerStoreId, $value, $
             ];
         }
 
-        $this->insertOnDuplicate('yotpo_customers_sync', $customersSyncData);
+        $this->insertOrUpdateCustomerSyncData($customersSyncData);
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php:425]: This could've been part of the commit declaring `insertOrUpdateCustomerSyncData`

**Review by LiorGingi — COMMENTED** [review]
This is up to commit 2f5abd038c5ff771dab5dfd8d95638581f7eb5a8.

File: Model/Sync/Customers/Processor.php [hunk 7]
```diff
@@ -184,25 +184,25 @@ public function processCustomer($customer, $customerAddress = null)
...
      * Process single customer entity
      *
-     * @param Customer $magentoCustomer
+     * @param Customer $customer
      * @param null|mixed $customerAddress
      * @return void
      * @throws NoSuchEntityException
      */
-    public function processSingleEntity($magentoCustomer, $customerAddress = null)
+    public function processSingleEntity($customer, $customerAddress = null)
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php:193]: customer what? entity? id? dto? array? object?

**Review by LiorGingi — CHANGES_REQUESTED** [review]
Up to 075deeb7b404a22b984082ab65477dcd14577c4a.

File: Model/Sync/Customers/Processor.php [hunk 8]
```diff
@@ -169,9 +169,10 @@ public function processCustomer($customer, $customerAddress = null)
             }
             $this->yotpoSmsBumpLogger->info(
                 __(
-                    'Process customer for the Magento Store ID: %1, Name: %2',
+                    'Starting syncing Customer to Yotpo - Magento Store ID: %1, Name: %2, CustomerID: %3',
```
Comment (LiorGingi) [inline Model/Sync/Customers/Processor.php:172]: **Customer ID**

**Review by LiorGingi — CHANGES_REQUESTED** [review]

File: Model/Sync/Customers/Main.php (hunk 1 above)
Comment (DanielxC) [reply Model/Sync/Customers/Main.php:100]: About first comment: 
Yes. its gets hash (array) and recreate the same hash (array) with the same keys.
About second comment:
Why not? insertOnDuplicate can insert array of items, but in this case its 1 item. Also it was passed as array before, it was just called finalData, the commit was taking unneeded recreation of array to shortcut of [$variable].

File: Model/Sync/Customers/Processor.php (hunk 2 above)
Comment (DanielxC) [reply Model/Sync/Customers/Processor.php:458]: There's already a function that prepare the data object, it is called few lines before it on each call:
https://github.com/YotpoLtd/magento2-module-messaging/blob/COR-1990-refactor-customers-sync-flows/Model/Sync/Customers/Processor.php#L284
https://github.com/YotpoLtd/magento2-module-messaging/blob/COR-1990-refactor-customers-sync-flows/Model/Sync/Customers/Processor.php#L203
I don't think we should separate it to methods.

Comment (DanielxC) [reply Model/Sync/Customers/Processor.php:458]: There's already a function that prepare the data object, it is called few lines before it on each call:
https://github.com/YotpoLtd/magento2-module-messaging/blob/COR-1990-refactor-customers-sync-flows/Model/Sync/Customers/Processor.php#L284
https://github.com/YotpoLtd/magento2-module-messaging/blob/COR-1990-refactor-customers-sync-flows/Model/Sync/Customers/Processor.php#L203
I will return the `$customerSyncData` and call `insertOrUpdateCustomerSyncData`.

**Review by LiorGingi — APPROVED** [review]

**Pull request closed without merging**